coding/src/reed_solomon/mod.rs

```diff
@@ -327,25 +355,38 @@ fn decode<H: Hasher>(
             .map_err(Error::ReedSolomon)?;
     }
     let encoding = encoder.encode().map_err(Error::ReedSolomon)?;
-    let recovery_shards: Vec<Vec<u8>> = encoding
-        .recovery_iter()
-        .map(|shard| shard.to_vec())
-        .collect();
+    let recovery_shards: Vec<&[u8]> = encoding.recovery_iter().collect();
     shards.extend(recovery_shards);
 
     // Build Merkle tree
     let mut builder = Builder::<H>::new(n);
-    let mut hasher = H::new();
-    for shard in &shards {
-        builder.add(&{
-            hasher.update(shard);
-            hasher.finalize()
+    if concurrency > 1 {
+        let pool = ThreadPoolBuilder::new()
+            .num_threads(concurrency)
+            .build()
+            .expect("unable to build thread pool");
+        let shard_hashes = pool.install(|| {
+            shards
+                .par_iter()
+                .map_init(H::new, |hasher, shard| {
+                    hasher.update(shard);
+                    hasher.finalize()
+                })
+                .collect::<Vec<_>>()
```

Comment:
do we need to `collect()` here? can we just populate the builder on the iterator?